### ISSUE TITLE
Access to low level ?

### ISSUE THREAD

**Issue by jonwright** [carsonswope/py-nvcomp#1]
For some LZ4 decompression I am calling "decompressStream" from the low level interface inside a cuda kernel. This could be due my incompetence in C++ (my code is [here](https://github.com/jonwright/bslz4decoders)). 

Are you planning to interface the low level parts too? I can't really see how my case could work without a python interface to set up the batching.

Obviously, it will be better for me if I can call your interface instead of copy+paste of the kernel code. Thanks for sharing your work!

**Comment by carsonswope** [comment]
Hi Jon,

I hadn't been planning to implement the bindings for the low-level interface but I'm certainly open to it.

It looks like wrote your own kernel that calls the device-side `decompressStream` function - if your kernel is significantly different from the global `lz4DecompressBatchKernel` kernel (which calls that function in the original nvcomp cuda file), than it seems python bindings will not be useful for you, because you are writing your own custom cuda kernels at that point.

On the other hand, if you could do what you need to with the (host-side) interfaces exposed by https://github.com/NVIDIA/nvcomp/blob/main/src/lowlevel/LZ4CompressionKernels.h, i.e. the `lz4BatchDecompress` function, then I'd be happy to implement those bindings.

**Comment by jonwright** [comment]
I will try to test it out and confirm via a little C program - it ought to be possible to go via the nvcomp interface.

**Comment by jonwright** [comment]
... the LZ4 part can indeed be done from the low level interface ([in C](https://github.com/jonwright/bslz4decoders/blob/main/bslz4decoders/test/lz4_via_nvcomp.cu)).

From python with ctypes I did not manage to figure this out, so a proper binding could help a lot. At the moment I am confused because there are a few unused arguments between the C interface and the kernel call:
```
void lz4BatchDecompress(    
     const uint8_t* const* device_in_ptrs,    
     const size_t* device_in_bytes,    
     const size_t* device_out_bytes,     // Not used
     const size_t batch_size,    
     void* temp_ptr,                     // Not used
     const size_t temp_bytes,            // Not used
     uint8_t* const* device_out_ptrs,    
     cudaStream_t stream);
```
I should be able to re-factor my code to call the same kernel instead, it just means setting up the pointers differently. I used offsets and these are absolutes:
```
lz4DecompressBatchKernel<<<grid, block, 0, stream>>>( 
        device_in_ptrs, 
        device_in_bytes, 
        batch_size, 
        device_out_ptrs);
```
It seems the `device_out_bytes` would be useful for bounds checks in some future version, even if I do not see what happens to it now. For the temp space I missed what is going on. Also what might happen if we feed it corrupted data. A spectacular crash perhaps?

Having the LZ4 batched compression option will be useful for other people I expect.

Do you want me to make an attempt in cython and send a pull request ? It would be my first attempt at writing cython code, so maybe not a great idea for something other people are going to use.

**Comment by carsonswope** [comment]
I've just pushed up some bindings for the batched lz4 API calls. (This project is my first cython code too - actually much easier than I expected.) Seems like it works in my basic test case - let me know if it's working for you!

And yeah, the unused arguments do make it confusing. I guess they are reserving all those arguments for future usage without breaking the API. Some kernels (like high-level CascadedCompressor) require some temp memory allocation, and looks like this one doesn't at the moment but maybe if they change the algorithm in the future it might. And `device_out_bytes` presumably would store 'how many bytes each chunk ended up decompressing to' - something the calling code currently would need to keep track of separately.

The question is of course whether to duplicate these unused arguments in the python bindings. I decided to remove them for now - not really sure what the 'right' thing to do is, probably a question for nvcomp team if/when I try to integrate this code into their repo.

(I noticed that the host side code will actually check to make sure `temp_ptr` and `device_out_bytes` are valid CUDA-accessible pointers, even though they are unused. Which is kind of annoying because it prevents you from just passing 0/nullptr to those arguments. That's the main reason I dropped the arguments from the python bindings for now)
